### ISSUE TITLE
Switching from another look-and-feel sometimes creates inconsistent combobox visuals

### ISSUE THREAD

**Issue by kirill-grouchnikov** [JFormDesigner/FlatLaf#17]
<img width="512" alt="image" src="https://user-images.githubusercontent.com/52874/67228533-6fd9bd80-f407-11e9-9565-4a56faa0840f.png">

See how the background behind the arrows is darker than the rest of the combobox fill.

```java
public class Test {
    public static void main(String[] args) {
        SwingUtilities.invokeLater(new Runnable() {
            @Override
            public void run() {
                try {
                    UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
                } catch (Throwable t) {
                    t.printStackTrace(System.err);
                }

                JComboBox comboHebrew = new JComboBox(new Object[] { "\u05e8\u05d0\u05e9\u05d9 1",
                        "\u05e8\u05d0\u05e9\u05d9 2", "\u05e8\u05d0\u05e9\u05d9 3",
                        "\u05e8\u05d0\u05e9\u05d9 4", "\u05e8\u05d0\u05e9\u05d9 5",
                        "\u05e8\u05d0\u05e9\u05d9 6", "\u05e8\u05d0\u05e9\u05d9 7",
                        "\u05e8\u05d0\u05e9\u05d9 8", "\u05e8\u05d0\u05e9\u05d9 9" });
                comboHebrew.setToolTipText("RTL combo");
                comboHebrew.setComponentOrientation(ComponentOrientation.RIGHT_TO_LEFT);
                comboHebrew.setMaximumRowCount(6);

                JComboBox comboRegular = new JComboBox(new Object[] { "item 1",
                        "item 2", "item 3",
                        "item 4", "item 5",
                        "item 6", "item 7",
                        "item 8", "item 9" });
                comboRegular.setToolTipText("LTR combo");
                comboRegular.setComponentOrientation(ComponentOrientation.LEFT_TO_RIGHT);
                comboRegular.setMaximumRowCount(6);

                JButton toFlat = new JButton("Set FlatDark!");
                toFlat.addActionListener(new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        SwingUtilities.invokeLater(new Runnable() {
                            @Override
                            public void run() {
                                try {
                                    UIManager.setLookAndFeel("com.formdev.flatlaf.FlatDarkLaf");
                                    for (Window w : Window.getWindows()) {
                                        SwingUtilities.updateComponentTreeUI(w);
                                    }
                                } catch (Throwable t) {
                                    t.printStackTrace(System.err);
                                }
                            }
                        });
                    }
                });

                JButton toNimbus = new JButton("Set Nimbus!");
                toNimbus.addActionListener(new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        SwingUtilities.invokeLater(new Runnable() {
                            @Override
                            public void run() {
                                try {
                                    UIManager.setLookAndFeel(new NimbusLookAndFeel());
                                    for (Window w : Window.getWindows()) {
                                        SwingUtilities.updateComponentTreeUI(w);
                                    }
                                } catch (Throwable t) {
                                    t.printStackTrace(System.err);
                                }
                            }
                        });
                    }
                });

                JButton toMetal = new JButton("Set Metal!");
                toMetal.addActionListener(new ActionListener() {
                    @Override
                    public void actionPerformed(ActionEvent e) {
                        SwingUtilities.invokeLater(new Runnable() {
                            @Override
                            public void run() {
                                try {
                                    UIManager.setLookAndFeel(new MetalLookAndFeel());
                                    for (Window w : Window.getWindows()) {
                                        SwingUtilities.updateComponentTreeUI(w);
                                    }
                                } catch (Throwable t) {
                                    t.printStackTrace(System.err);
                                }
                            }
                        });
                    }
                });

                JFrame frame = new JFrame("Combo test");
                frame.setLayout(new BorderLayout());

                JPanel flow = new JPanel(new FlowLayout());
                comboRegular.setPreferredSize(new Dimension(comboRegular.getPreferredSize().width + 50,
                        comboRegular.getPreferredSize().height));
                flow.add(comboRegular);
                comboHebrew.setPreferredSize(new Dimension(comboHebrew.getPreferredSize().width + 50,
                        comboHebrew.getPreferredSize().height));
                flow.add(comboHebrew);
                frame.add(flow, BorderLayout.NORTH);

                JPanel flowButtons = new JPanel(new FlowLayout());
                flowButtons.add(toNimbus);
                flowButtons.add(toMetal);
                flowButtons.add(toFlat);
                frame.add(flowButtons, BorderLayout.SOUTH);

                frame.setSize(400, 400);
                frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
                frame.setLocationRelativeTo(null);
                frame.setVisible(true);
            }
        });
    }
}
```

**Comment by DevCharly** [comment]
Good catch 👍 

The problem is that the `AquaComboBoxUI` changes `opaque` to `false` and [does not restore it correctly](https://github.com/openjdk/jdk/blob/6bab0f539fba8fb441697846347597b4a0ade428/src/java.desktop/macosx/classes/com/apple/laf/AquaComboBoxUI.java#L52-L67). And `FlatComboBoxUI` does not paint correctly if opaque is false...

It is always a problem when a LaF changes something in the component and does not restore it correctly.

There is a similar problem with Windows LaF (in Java 9 and later), that [sets a non-UIResouce EmptyBorder](https://github.com/openjdk/jdk/blob/0d5cbcda529a71b6f5dacb2a98c1708588288f5c/src/java.desktop/windows/classes/com/sun/java/swing/plaf/windows/WindowsComboBoxUI.java#L157) at the combobox. When switching from Windows LaF to FlatLaf the combobox borders are gone:

![image](https://user-images.githubusercontent.com/5604048/67287783-04adeb00-f4dc-11e9-83cc-1e740233f883.png)

Probably have to remember what LaF was active before switching to FlatLaf to fix those issues...

**Comment by kirill-grouchnikov** [comment]
I'm guilty of not restoring that in Substance as well (for combo boxes). I do it for some other controls (like buttons), storing the original opacity when the UI delegate is installed, and resetting it when it is uninstalled.

The thing I don't like about this is what happens when the application code calls setOpacity in the middle of the app "flow". I have an explicit FAQ entry saying that apps shouldn't do that. Probably I should restore the opacity bit for comboboxes.

**Comment by DevCharly** [comment]
For the opaque property, the solution is to use following in `installDefaults()` (after invoking super method):
~~~java
LookAndFeel.installProperty( comboBox, "opaque", Boolean.FALSE);
~~~
(change last parameter to `TRUE` if needed).

This invokes `JComponent.setOpaque(...)` only if it was not set explicitly.

But this works only correctly if the BasicXyzUI (or all other LaFs) also invokes this method, to reset opaque when switching to another LaF. `BasicComboBoxUI`, `BasicButtonUI` and many other Basic UIs use this method.

**Comment by kirill-grouchnikov** [comment]
`Convenience method for installing a property with the specified name and value on a component if that property has not already been set by the developer`

Unfortunately with how Substance paints the control visuals (rounded corners, partial translucency in some cases), I have to set opacity to false no matter what the application "preference" is.

**Comment by DevCharly** [comment]
I see. Because of the round corners?

In FlatLaf are also round corners. And "focused" borders outside of the components (in Darcula and IntelliJ themes). But the components are opaque.

![image](https://user-images.githubusercontent.com/5604048/67301029-2795ca80-f4ef-11e9-8aab-5424b33e4c2c.png)

The trick I use (borrowed from IDEA) is to fill the background of the component with the the background color of a non-opaque parent.

https://github.com/JFormDesigner/FlatLaf/blob/master/flatlaf-core/src/main/java/com/formdev/flatlaf/ui/FlatUIUtils.java#L136-L157

Works fine as long as the panel uses a solid color. Will fail for gradient or image backgrounds...

**Comment by kirill-grouchnikov** [comment]
Right, my preference is to say that Substance doesn't play well with application calls to `setOpacity`. It can get pretty complicated if the child component needs to paint the parent's background with nested multi-colored panels, toolbars etc.

**Comment by DevCharly** [comment]
fixed in master

`LookAndFeel.installProperty( comboBox, "opaque", Boolean.FALSE);` works well for me and the components now even paint well if opaque was explicitly set to true. But it is important to [fill the whole background](https://github.com/JFormDesigner/FlatLaf/blob/bc10c4e871b51bee93f31c32777c9bea34fa018f/flatlaf-core/src/main/java/com/formdev/flatlaf/ui/FlatComboBoxUI.java#L276-L279) in `MyUI.update(Graphics g, JComponent c)` somehow if opaque is true. Otherwise I had some garbage (if single components are repainted).

**Comment by JFormDesigner** [comment]
fixed in 0.17
